### ISSUE TITLE
IllegalAccessError: no such constructor: net.fortuna.ical4j.util.JCacheTimeZoneCache.<init>()void/newInvokeSpecial since upgrade to 3.0.0

### ISSUE THREAD

**Issue by malkusch** [ical4j/ical4j#244]
I did the recent upgrade to 3.0.0 and I'm running on Java-8 still. Since that upgrade I have again a JCache ClassNotFound issue, but this time I do have set the property net.fortuna.ical4j.timezone.cache.impl: net.fortuna.ical4j.util.MapTimeZoneCache. That is the relevant stack trace:
```
Caused by: java.lang.BootstrapMethodError: java.lang.IllegalAccessError: no such constructor: net.fortuna.ical4j.util.JCacheTimeZoneCache.<init>()void/newInvokeSpecial
	at net.fortuna.ical4j.model.TimeZoneLoader.cacheInit(TimeZoneLoader.java:275) ~[ical4j-3.0.0.jar:na]
	at net.fortuna.ical4j.model.TimeZoneLoader.<init>(TimeZoneLoader.java:81) ~[ical4j-3.0.0.jar:na]
	at net.fortuna.ical4j.model.TimeZoneRegistryImpl.<init>(TimeZoneRegistryImpl.java:125) ~[ical4j-3.0.0.jar:na]
	at net.fortuna.ical4j.model.TimeZoneRegistryImpl.<init>(TimeZoneRegistryImpl.java:116) ~[ical4j-3.0.0.jar:na]
	at net.fortuna.ical4j.model.DefaultTimeZoneRegistryFactory.createRegistry(DefaultTimeZoneRegistryFactory.java:48) ~[ical4j-3.0.0.jar:na]
	at net.fortuna.ical4j.data.CalendarBuilder.<init>(CalendarBuilder.java:105) ~[ical4j-3.0.0.jar:na]
	at de.malkusch.trashcollection.infrastructure.schedule.ical.VEventRepository.downloadVEvents(VEventRepository.java:46) ~[classes/:na]
	at de.malkusch.trashcollection.infrastructure.schedule.ical.VEventRepository.<init>(VEventRepository.java:35) ~[classes/:na]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_172]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_172]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_172]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[na:1.8.0_172]
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:170) ~[spring-beans-5.0.7.RELEASE.jar:5.0.7.RELEASE]
	... 80 common frames omitted
Caused by: java.lang.IllegalAccessError: no such constructor: net.fortuna.ical4j.util.JCacheTimeZoneCache.<init>()void/newInvokeSpecial
	at java.lang.invoke.MethodHandleNatives.linkMethodHandleConstant(MethodHandleNatives.java:483) ~[na:1.8.0_172]
	... 93 common frames omitted
Caused by: java.lang.NoClassDefFoundError: javax/cache/configuration/Configuration
	at java.lang.invoke.MethodHandleNatives.resolve(Native Method) ~[na:1.8.0_172]
	at java.lang.invoke.MemberName$Factory.resolve(MemberName.java:975) ~[na:1.8.0_172]
	at java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1000) ~[na:1.8.0_172]
	at java.lang.invoke.MethodHandles$Lookup.resolveOrFail(MethodHandles.java:1394) ~[na:1.8.0_172]
	at java.lang.invoke.MethodHandles$Lookup.linkMethodHandleConstant(MethodHandles.java:1750) ~[na:1.8.0_172]
	at java.lang.invoke.MethodHandleNatives.linkMethodHandleConstant(MethodHandleNatives.java:477) ~[na:1.8.0_172]
	... 93 common frames omitted
Caused by: java.lang.ClassNotFoundException: javax.cache.configuration.Configuration
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381) ~[na:1.8.0_172]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_172]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349) ~[na:1.8.0_172]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_172]
	... 99 common frames omitted
```
The weird thing is, when I'm debugging the relevant code. I see that the offending Optional in TimeZoneLoader.java:275 has in fact an instance of MapTimeZoneCache. So it should never come to the orElseGet() part. However it seems that Java-8 is not happy with the non existing JCache from that constructor. Does it work for you when you have no JCache on the class path?

As workaround putting the JCache-API into the class path fixes that issue for me:
```
<dependency>
  <groupId>javax.cache</groupId>
  <artifactId>cache-api</artifactId>
</dependency>
```

**Comment by benfortuna** [comment]
Thanks for the bug report. I've put a fix in now and will release shortly.

**Comment by malkusch** [comment]
Thanks for the quick fix. I am really curious about this one as I lack some understanding in this space. Do you have any further knowledge why this issue happens at all and also why the fix works? I saw your commit and I would consider both identical.

**Comment by benfortuna** [comment]
I'm a bit new to method references (i.e. JCacheTimeZoneCache::new), but it seems that it was invoking the contructor regardless.

Whereas the fix uses a lambda expression as a shorthand for the following:

```
new Supplier<TimeZoneCache>() {
            @Override
            public TimeZoneCache get() {
                return new JCacheTimeZoneCache();
            }
        }
```

which will only invoke the constructor if the configured cache is null.

**Comment by malkusch** [comment]
>  but it seems that it was invoking the contructor regardless.

No, that shouldn't be the case. It seems more that there's an implementation detail leaking as a result from different implementations between `Foo::new` and `() -> new Foo()`. I posted a [question at StackOverflow](https://stackoverflow.com/questions/51113252/whats-the-difference-between-foonew-and-new-foo) to gain more insights. Consent is not there yet, but if we are in the country of implementation details, I would suggest the current fix is not reliable and suggest to add the JCache-API as dependency.

**Comment by benfortuna** [comment]
Ok thanks for the reference.

 I had a look at the stack overflow thread, and if consensus is the fix isn't reliable I will revert to original code that doesn't rely on lambdas or method references (ie. The code snippet I posted in the last comment)

**Comment by malkusch** [comment]
Consent is implementation detail and might change in the future. I think a better fix would be just adding the JCache-API as dependency:
```
<dependency>
    <groupId>javax.cache</groupId>
    <artifactId>cache-api</artifactId>
    <version>1.1.0</version>
</dependency>
```

**Comment by stellingsimon** [comment]
@malkusch I would strongly favor an optional dependency over a hard one. 

A proper, stable fix is to simply check the classpath before invoking JCache, and there is no requirement to force users of the library into including this dependency in the classpath if they don't want to use JCache from within ical4j. (I don't.)

For a more general discussion of why minimizing dependencies brings more value to the libraries user's, see this blog post by Lukas Eder (a "Java Champion"): https://blog.jooq.org/tag/dependency-hell/

**Comment by malkusch** [comment]
@stellingsimon How would you then fix this issue?

**Comment by benfortuna** [comment]
Ok so I've updated the code to use reflection to load the default cache implementation. 

Would appreciate if you can have a look at the code in develop branch and let me know of any further issues with the fix.

**Comment by benfortuna** [comment]
Fix released in ical4j 3.0.1

**Comment by benfortuna** [comment]
Fix applied. Closing issue (reopen if still concerns).
